lib/hsa/hsaRfiUpdateBadInputs.cpp

```diff
@@ -51,30 +49,59 @@ hsaRfiUpdateBadInputs::~hsaRfiUpdateBadInputs() {
 int hsaRfiUpdateBadInputs::wait_on_precondition(int gpu_frame_id) {
     (void)gpu_frame_id;
 
-    uint8_t* frame =
-        wait_for_full_frame(_network_buf, unique_name.c_str(), _network_buf_precondition_id);
-    if (frame == nullptr)
-        return -1;
-
-    _network_buf_precondition_id = (_network_buf_precondition_id + 1) % _network_buf->num_frames;
+    // Check for bad input updates
+    std::lock_guard<std::mutex> lock(update_mutex);
+    if (first_pass) {
+        uint8_t* frame = wait_for_full_frame(_in_buf, unique_name.c_str(), _in_buf_precondition_id);
+        if (frame == NULL)
+            return -1;
+        first_pass = false;
+        frames_to_update = device.get_gpu_buffer_depth();
+        memcpy(host_mask, frame, input_mask_len);
+        num_bad_inputs = get_rfi_num_bad_inputs(_in_buf, _in_buf_precondition_id);
+        DEBUG("finalize_frame for gpu_frame_id={:d} using _in_buf_precondition_id={:d}",
+              gpu_frame_id, _in_buf_precondition_id);
+        mark_frame_empty(_in_buf, unique_name.c_str(), _in_buf_precondition_id);
+        _in_buf_precondition_id = (_in_buf_precondition_id + 1) % _in_buf->num_frames;
```

Comment:
Factor out this and lines 76-82 to an inline function maybe?  The only local variable is `gpu_frame_id` by the looks of it.   Also the debug message still talks about `finalize_frame` which is no longer true.